File: node/day2part2.js
#!/usr/bin/env node
/*
 * A Javascript / NodeJS implementation of AOC 2021 Day 2 Part 2
 *
 * Mainly to prove I can.
 *
 * Algorithm is a direct clone of that used in python/day2part2.py so for
 * explanations see there.
 */

fs = require('fs');

//const inputfile = "data/day2test.txt";
const inputfile = "data/day2part1.txt";

const commands  = fs.readFileSync(inputfile,'UTF-8').split(/\r?\n/);

cmdCount = 0;
horizontal = 0;
aim = 0;
depth = 0;

commands.forEach(command => {
  cmdCount++;
  const cmd = command.split(" ",2);
  const size = Number(cmd[1]);
  switch(cmd[0].toLowerCase()) {
    case "forward" :
      horizontal += size;
      depth += (aim * size);
      if(depth < 0) { depth = 0};
      break;
    case "down" :
      aim += size;
      break;
    case "up" :
      aim -= size;
      break;
    default :
      //console.log("Unable to parse command " + command + " -  I got cmd=" + cmd[0] + ", size=" + size);
      cmdCount -= 1;
  }
});

const positionProduct = horizontal * depth;
console.log("Final position after " + cmdCount + " valid commands is horizontal = " + horizontal + ", depth = " + depth);
console.log("Position Product is therefore: " + positionProduct);

File: python/day2part2.py
#!/usr/bin/python3
# Python solution for AOC 2021 Day 2, Part 2
# Given an input of steering commands, work out the position of the submarine
# permuted by something called "aim"

#inputfile="data/day2test.txt"
inputfile="data/day2part1.txt"

# horizontal is "pure" (affected directly by forward/back commands)
# but depth is affected by aim and horizontal commands.
# essentially up/down commands change the angle of steering fins...
horizontal = 0
aim = 0
depth = 0

cmdcount=0
with open(inputfile,'r') as instructions:
    for command in instructions:
        cmdcount+=1
        #split into a <direction> <distance> pair
        cmd,size = command.split(' ')
        size = int(size)
        #Action based on command:
        if cmd.lower() == "forward" :
            horizontal += size
            depth += (aim * size)
            if depth < 0 :
                depth = 0 # Guard against flying out of the ocean....
        elif cmd.lower() == "down" :
            aim += size
        elif cmd.lower() == "up" :
            aim -= size
        else :
            print(f"FAILED TO PARSE COMMAND: {command} - I only got {cmd},{distance}")
            cmdcount-=1

print(f"Final position after {cmdcount} valid commands is horizontal = {horizontal}, depth = {depth}")
print(f"Position Product is therefore: {horizontal * depth}")
